runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -745,6 +1091,295 @@ JITServerAOTCache::printStats(FILE *f) const
+// incompatible with the running server.
+JITServerAOTCache *
+JITServerAOTCache::readCache(FILE *f, const std::string &name, TR_Memory &trMemory)
+   {
+   if (!JITServerAOTCacheMap::cacheHasSpace())
+      return NULL;
+
+   JITServerAOTCacheHeader header = {0};
+   if (1 != fread(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return NULL;
+
+   if (!isCompatibleSnapshotVersion(header._version))
+      return NULL;
+
+   JITServerAOTCache *cache = NULL;
+   try
+      {
+      cache = new (TR::Compiler->persistentGlobalMemory()) JITServerAOTCache(name);
+      }
+   catch (const std::exception &e)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         {
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache allocation failed with exception: %s", e.what());
+         }
+      }
+
+   if (!cache)
+      return NULL;
+
+   bool readSuccess = false;
+   try
+      {
+      readSuccess = cache->readCache(f, header, trMemory);
+      }
+   catch (const std::exception &e)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         {
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache reading failed with exception: %s", e.what());
+         }
+      }
+
+   if (!readSuccess)
+      {
+      cache->~JITServerAOTCache();
+      TR::Compiler->persistentGlobalMemory()->freePersistentMemory(cache);
+      cache = NULL;
+      }
+
+   return cache;
+   }
+
+template<typename K, typename V, typename H> bool
+JITServerAOTCache::readRecords(FILE *f,
+                               size_t numRecordsToRead,
+                               PersistentUnorderedMap<K, V *, H> &map,
+                               V *&traversalHead,
+                               V *&traversalTail,
+                               Vector<V *> &records,
+                               const Vector<AOTCacheClassLoaderRecord *> &classLoaderRecords,
+                               const Vector<AOTCacheClassRecord *> &classRecords,
+                               const Vector<AOTCacheMethodRecord *> &methodRecords,
+                               const Vector<AOTCacheClassChainRecord *> &classChainRecords,
+                               const Vector<AOTCacheWellKnownClassesRecord *> &wellKnownClassesRecords,
+                               const Vector<AOTCacheAOTHeaderRecord *> &aotHeaderRecords)
+   {
+   for (size_t i = 0; i < numRecordsToRead; ++i)
+      {
+      if (!JITServerAOTCacheMap::cacheHasSpace())
+         return false;
+
+      V *record = V::read(f, classLoaderRecords, classRecords, methodRecords, classChainRecords, wellKnownClassesRecords, aotHeaderRecords);
+      if (!record)
+         return false;
+
+      if ((record->data().id() >= records.size() ||
+          records[record->data().id()]))
+         {
+         AOTCacheRecord::free(record);
+         return false;
+         }
+
+      addToMap(map, traversalHead, traversalTail, key(*record), record);
```

Comment:
Should we also check that the record keys are unique? E.g. `addToMap()` could return `false` if the key already exists.